### DISCUSSION TITLE
Dealing with Duplicate Data in Columns with Unique Constraints

### DISCUSSION THREAD

**Discussion by sule26** [sqlalchemy/sqlalchemy#11126] [Usage Questions] [answered]
I'm encountering an issue while using the faker library to populate tables in my project. Specifically, when I utilize session.add_all to insert a large number of rows (e.g., 1000), I intermittently encounter errors indicating attempted insertion of values already present in the database.

To mitigate this, I'm considering a process where I use session.add within a loop to check for existing data before insertion. However, this approach seems inefficient, as it involves multiple database checks and modifications.

I'm seeking advice on a more performant solution to handle duplicate data when inserting rows with unique constraints. How can I optimize this process to ensure efficient insertion of data without encountering duplication errors?

Your insights and suggestions would be greatly appreciated. Thank you!

**Comment by zzzeek** [comment, marked as answer]
I have my own approach to this which I think is best, however, it tends to not score a lot of points for slickness , because it's not trying to be slick, it's trying to do the job well.

the approach is:

1. partition your data into batches (1000 items is actualyl a good batch size)
2. pre-SELECT all rows that overlap with those 1000 items.   This could be just selecting the primary key alone, or whatever you need to fetch that identifies the row already exists.   the query here may be one of two forms:
a.  the SELECT uses an IN clause that has all the primary keys of the 1000 items in your batch.  or
b. the SELECT selects from a range that refers to the start and end of your 1000 item batch.  doing it this way requires that the batches are ordered.
3. do an INSERT for all the entries in the batch that werent in what you SELECTed.

This is really the same idea as you have, the "loop where you check for existing data", but do it in batches.  Check for 1000 records at once.
